### PR TITLE
Simplistic pipeline version

### PR DESCRIPTION
**Pull request by getand**
I've changes the implementation here so that it's possible to use the pipeline without the need of implementing steps and thus having only one directional dependencies. Let's discuss monday.

Probably a bit shitty to review here.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,93 +22,114 @@ Or install it yourself as:
...
+by the step yielded to the constructor. 
 
 ```ruby
-class MyPipeline < NxtPipeline::Pipeline
-  pipe_attr :words
+pipeline = NxtPipeline::Pipeline.new do |p|
+  # Add a named constructor that will be used to execute your steps later
+  # All options that you pass in your step will be available through accessors in your constructor 
+  p.constructor(:service, default: true) do |step, arg|
+    step.service_class.new(optiopns: arg).call
```
Comment (nsommer) [inline README.md]:
```suggestion
    step.service_class.new(options: arg).call
```

File: lib/nxt_pipeline/error_callback.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+      @block = block
+    end
+
+    attr_accessor :errors, :block
+
+    def applies_to_error?(error)
+      (error.class.ancestors & errors).any?
+    end
+
+    def call(step, arg, error)
```
Comment (nsommer) [inline lib/nxt_pipeline/error_callback.rb]: In general I like to have `&block` as an explicit argument in the method definition.

```suggestion
    def call(step, arg, error, &block)
```

File: lib/nxt_pipeline/pipeline.rb [hunk 3]
```diff
@@ -1,81 +1,97 @@
...
+      registry[name] = constructor
+
+      return unless opts.fetch(:default, false)
+      default_constructor ? (raise ArgumentError, 'Default step already defined') : self.default_constructor = constructor
     end
 
-    def failed?
-      failed_step.present?
+    def step(type = nil, **opts, &block)
+      constructor = if block_given?
```
Comment (nsommer) [inline lib/nxt_pipeline/pipeline.rb]: Indentation seems off for two spaces until L41.

File: lib/nxt_pipeline/pipeline.rb [hunk 4]
```diff
@@ -1,81 +1,97 @@
...
-      @pipe_attr = attrs.first.fetch(self.class.pipe_attr_name)
-    end
+    def execute_step(step, arg)
+      self.current_step = step.to_s
+      result = step.execute(arg)
 
-    def handle_segment_burst(error, step)
-      @failed_step = step.name.split('::').last.underscore
+      if result # step was successful
+        log[current_step] = 'success'
```
Comment (nsommer) [inline lib/nxt_pipeline/pipeline.rb]: Maybe it makes sense to use symbols instead of strings?

File: lib/nxt_pipeline/step.rb [hunk 5]
```diff
@@ -1,47 +1,32 @@
...
+    attr_reader :opts
 
-    def validate_initialize_args(*args)
-      raise ArgumentError, arguments_missing_msg(self.step_args) if args.empty?
-
-      keyword_args = args.first
-      missing_keyword_args = self.step_args.reject do |arg|
-        keyword_args.include?(arg)
+    def define_attr_readers(opts)
+      opts.each do |k,v|
```
Comment (nsommer) [inline lib/nxt_pipeline/step.rb]:
```suggestion
      opts.each do |k, v|
```

Comment (nsommer) [reply lib/nxt_pipeline/step.rb]:
```suggestion
      opts.each do |key, value|
```

File: lib/nxt_pipeline/error_callback.rb (hunk 2 above)
Comment (getand) [reply lib/nxt_pipeline/error_callback.rb]: Totally agree. Usually always do that. I hate when methods take a block but don`t have it in the signature.

File: lib/nxt_pipeline/pipeline.rb (hunk 4 above)
Comment (getand) [reply lib/nxt_pipeline/pipeline.rb]: I mean it's ment to be readable. That'S the only thing I care about at the moment.

Comment (rkallensee) [reply lib/nxt_pipeline/pipeline.rb]: I agree, I would also prefer to use symbols.

Comment (getand) [reply lib/nxt_pipeline/pipeline.rb]: @rkallensee @nsommer But it does not make sense to use symbols. It's a human description. When there is an error it contains a sentence.

**Pull request merged**